src/cairo/bootloader.md

```diff
@@ -1 +1,199 @@
+
+### Builtins
+
+Builtins are pre-allocated memory cells that can be used for certain commonly-used “builtin” functions. As of Cairo v0.13.1, these builtins are currently supported:
+
+- Output
+- Pedersen
+- Range check
+- Bitwise
+- Elliptic curve operation
+- Keccak
+- Poseidon
+- Range check (96 bits)
+
+Each builtin will have a separate segment in the memory vector and depending on the builtin, you can just input the values you want to run the builtin on or additionally use the output of the builtin.
+
+For example, to use the range-check builtin you can just input the values you want to range-check in a certain memory segment and the CairoVM will create the necessary constraints for you.
+
+For a Pedersen builtin, on the other hand, for each builtin use, you can input two values and read the next cell as output since the VM provides a guarantee that the next cell will contain the correct output (i.e. the Pedersen hash of the two input values).
+
+<div style="text-align: center;">
+    <img src="builtin_memory_layout.png" alt="Builtin memory layout" width="80%">
+</div>
+
+In order to run a program that uses one of these builtins, we need to specify a specific layout to CairoVM that supports the builtins that the program uses. Below is a subset of layouts that are currently supported by the CairoVM. (Check this [code](https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/types/instance_definitions/builtins_instance_def.rs) for the updated, comprehensive list)
+
+|             | small | recursive | dex | recursive_with_poseidon | starknet | starknet_with_keccak |
+| ----------- | :---: | :-------: | :-: | :---------------------: | :------: | :------------------: |
+| output      |   O   |     O     |  O  |            O            |    O     |          O           |
+| pedersen    |   O   |     O     |  O  |            O            |    O     |          O           |
+| range_check |   O   |     O     |  O  |            O            |    O     |          O           |
+| bitwise     |       |     O     |     |            O            |    O     |          O           |
+| ecdsa       |       |           |  O  |                         |    O     |          O           |
+| poseidon    |       |           |     |            O            |    O     |          O           |
+| ec_op       |       |           |     |                         |    O     |          O           |
+| keccak      |       |           |     |                         |          |          O           |
+
+### Hints
+
+Hints are non-deterministic computation (i.e. code that is run outside of the VM to populate values that are validated inside the VM)
+
+For example, when computing the square root of an integer in a prime field, directly computing this using constraints is expensive. So instead, we can use hint to calculate the square root and create a constraint inside the VM that checks that the power of the given square root is equal to the original value.
```

Comment:
Is the square root computed in the field or over the integers?

Reply:
I think over the integers? Is there a case where it should be computed in the field?

Reply:
not all numbers in a field necessarily have a sqrt (https://en.wikipedia.org/wiki/Euler%27s_criterion), you have the same issue if you give the example of division (what is 4 divided by 3 in a field? it's going to be 4 * inverse_of_3 mod p)